### PR TITLE
Implement most of bulk memory

### PR DESCRIPTION
**Pull request by fitzgen**
Felt like a good time to start landing some of this work :)

### CODE REVIEW

File: crates/runtime/src/instance.rs [hunk 1]
```diff
@@ -1174,11 +1181,29 @@ fn initialize_tables(instance: &Instance) -> Result<(), InstantiationError> {
...
+                "active element segment does not fit in table".into(),
+            ));
+        }
+
         for (i, func_idx) in init.elements.iter().enumerate() {
             let anyfunc = instance.get_caller_checked_anyfunc(*func_idx);
             table
                 .set(u32::try_from(start + i).unwrap(), anyfunc)
-                .unwrap();
+                // Note that when multi-value is disabled, this will never fail
```
Comment (fitzgen) [inline crates/runtime/src/instance.rs]: TODO for myself: s/multi-value/bulk-memory/

File: crates/runtime/src/instance.rs [hunk 2]
```diff
@@ -1174,11 +1181,29 @@ fn initialize_tables(instance: &Instance) -> Result<(), InstantiationError> {
...
+                // since we bounds check table element initialization before
+                // doing any table slot writes. However when multi-value is
```
Comment (fitzgen) [inline crates/runtime/src/instance.rs]: ditto

File: crates/runtime/src/instance.rs [hunk 3]
```diff
@@ -523,6 +538,243 @@ impl Instance {
...
+    /// passive element.
+    ///
+    /// # Errors
+    ///
+    /// Returns a `Trap` error when the range within the table is out of bounds
+    /// or the range within the passive element is out of bounds.
+    pub(crate) fn table_init(
+        &self,
+        table_index: TableIndex,
+        elem_index: PassiveElemIndex,
```
Comment (alexcrichton) [inline crates/runtime/src/instance.rs]: I've seen this a few times now but I've got a question now at this point, why `PassiveElemIndex`? I thought that the index space included all elements, right? You can `table.init` from an active segment I thought?

File: crates/runtime/src/instance.rs [hunk 4]
```diff
@@ -523,6 +538,243 @@ impl Instance {
...
+                },
+                backtrace: Backtrace::new(),
+            });
+        }
+
+        for (dst, src) in (dst..dst + len).zip(src..src + len) {
+            table
+                .set(dst, elem[src as usize].clone())
+                .expect("should never panic because we already did the bounds check above");
+        }
```
Comment (alexcrichton) [inline crates/runtime/src/instance.rs]: Mind throwing a TODO here to lower this to a `memcpy`? I think we should be able to do that, right?

File: crates/runtime/src/instance.rs [hunk 5]
```diff
@@ -523,6 +538,243 @@ impl Instance {
...
+        Ok(())
+    }
+
+    /// Drop an element.
+    pub(crate) fn elem_drop(&self, elem_index: PassiveElemIndex) {
+        // https://webassembly.github.io/reference-types/core/exec/instructions.html#exec-elem-drop
+
+        let mut passive_elements = self.passive_elements.borrow_mut();
+        // Note that dropping a non-passive element is a no-op (not a trap).
+        if let Some(elem) = passive_elements.get_mut(&elem_index) {
```
Comment (alexcrichton) [inline crates/runtime/src/instance.rs]: Instead of `get_mut` should this use `remove` to remove it entirely? (and then we treat non-present elements as empty)

File: crates/runtime/src/instance.rs [hunk 6]
```diff
@@ -523,6 +538,243 @@ impl Instance {
...
+            });
+        }
+
+        let dst = isize::try_from(dst).unwrap();
+        let src = isize::try_from(src).unwrap();
+
+        // Bounds and casts are checked above, by this point we know that
+        // everything is safe.
+        unsafe {
+            let dst = memory.base.offset(dst);
```
Comment (alexcrichton) [inline crates/runtime/src/instance.rs]: FWIW you can use `add` here to use the `usize` type rather than `isize`

File: crates/runtime/src/instance.rs [hunk 7]
```diff
@@ -523,6 +538,243 @@ impl Instance {
...
+            .checked_add(len)
+            .map_or(true, |m| m as usize > memory.current_length)
+        {
+            return Err(Trap::Wasm {
+                desc: TrapDescription {
+                    source_loc,
+                    trap_code: ir::TrapCode::HeapOutOfBounds,
+                },
+                backtrace: Backtrace::new(),
+            });
```
Comment (alexcrichton) [inline crates/runtime/src/instance.rs]: We may want to add a method to `Trap` perhaps like:

```rust
Trap::wasm(source_loc, ir::TrapCode::HeapOutOfBounds)
```

to make this a bit less wordy

File: crates/runtime/src/instance.rs [hunk 8]
```diff
@@ -523,6 +538,243 @@ impl Instance {
...
+            let foreign_index = foreign_instance.memory_index(foreign_memory);
+            foreign_instance.defined_memory_fill(foreign_index, dst, val, len, source_loc)
+        }
+    }
+
+    /// Get a table by index regardless of whether it is locally-defined or an
+    /// imported, foreign table.
+    pub(crate) fn get_table(&self, table_index: TableIndex) -> &Table {
+        if let Some(defined_table_index) = self.module.defined_table_index(table_index) {
+            &self.tables[defined_table_index]
```
Comment (alexcrichton) [inline crates/runtime/src/instance.rs]: Perhaps this could delegate to `get_defined_table` for symmetry?

File: crates/runtime/src/instance.rs [hunk 9]
```diff
@@ -648,12 +902,18 @@ impl InstanceHandle {
...
         );
 
-        // Check initializer bounds before initializing anything.
-        check_table_init_bounds(instance)?;
-        check_memory_init_bounds(instance, data_initializers)?;
+        // Check initializer bounds before initializing anything. Only do this
+        // when bulk memory is disabled, since the bulk memory proposal changes
+        // instantiation such that the intermediate results of failed
+        // initializations are visible.
+        if !is_bulk_memory {
```
Comment (alexcrichton) [inline crates/runtime/src/instance.rs]: Oh bummer :(

I thought that the bulk memory proposal just changed types of errors, not changing semantics like this. Passing around this `is_bulk_memory` boolean is sort of a bummer, but given this change in semantics I don't think we can do much about it.

File: crates/api/src/instance.rs [hunk 10]
```diff
@@ -19,16 +19,29 @@ impl Resolver for SimpleResolver<'_> {
...
+                config.validating_config.operator_config.enable_bulk_memory,
+                &mut resolver,
+            )
             .map_err(|e| -> Error {
                 match e {
                     InstantiationError::StartTrap(trap) => Trap::from_jit(trap).into(),
+                    e @ InstantiationError::TableOutOfBounds(_)
+                    | e @ InstantiationError::MemoryOutOfBounds(_) => {
+                        let msg = e.to_string();
+                        if config.validating_config.operator_config.enable_bulk_memory {
```
Comment (alexcrichton) [inline crates/api/src/instance.rs]: Since we're passing this boolean into the instantiation anyway, should this sort of error handling happen in the instance? I was hoping we'd be able to avoid passing the boolean all the way down, but given the change in operational semantics I don't think we'll be able to change that.

File: crates/runtime/src/instance.rs [hunk 11]
```diff
@@ -978,8 +1259,25 @@ fn initialize_memories(
...
+            let to_init = &mut mem_slice[start..actual_end];
+            to_init.copy_from_slice(&init.data[..init.data.len() - num_uncopied]);
+
+            // Note that if bulk memory is disabled, we'll never hit this case
+            // because we did an eager bounds check in
+            // `check_memory_init_bounds`. However when bulk memory is enabled,
+            // we need the incremental writes up to the OOB trap to be visible,
+            // since this is dictated by the updated spec for the bulk memory
+            // proposal.
+            if num_uncopied > 0 {
```
Comment (alexcrichton) [inline crates/runtime/src/instance.rs]: Wait now I think I'm a bit confused, I thought the bulk memory proposal did the bounds checks up-front nowadays? For example the `memory.copy` implementation up above did the bounds check before copying any memory.

File: crates/runtime/src/libcalls.rs [hunk 12]
```diff
@@ -137,3 +142,195 @@ pub unsafe extern "C" fn wasmtime_imported_memory32_size(
 
     instance.imported_memory_size(memory_index)
 }
+
+/// Implementation of `table.copy` when both tables are locally defined.
+#[no_mangle]
+pub unsafe extern "C" fn wasmtime_table_copy_defined_defined(
```
Comment (alexcrichton) [inline crates/runtime/src/libcalls.rs]: While I realize that you're just persisting the same style as other functions in this module, I think the `#[no_mangle]` here isn't necessary?

Additionally, how come there are 4 table copy functions (permutations of imported/local src/dst) but for memory.copy there's only one? It looks like each of these functions may save one branch when figuring out the table based on the index, but that seems pretty cheap compared to the copy itself.

(seemed a bit wordy to have 4 functions)

File: crates/runtime/src/libcalls.rs [hunk 13]
```diff
@@ -137,3 +142,195 @@ pub unsafe extern "C" fn wasmtime_imported_memory32_size(
...
+    source_loc: u32,
+) {
+    let dst_table_index = DefinedTableIndex::from_u32(dst_table_index);
+    let src_table_index = DefinedTableIndex::from_u32(src_table_index);
+    let source_loc = ir::SourceLoc::new(source_loc);
+    let instance = (&mut *vmctx).instance();
+    let dst_table = instance.get_defined_table(dst_table_index);
+    let src_table = instance.get_defined_table(src_table_index);
+    if let Err(trap) = Table::copy(dst_table, src_table, dst, src, len, source_loc) {
+        raise_lib_trap(trap);
```
Comment (alexcrichton) [inline crates/runtime/src/libcalls.rs]: This ends up not running destructors and is pretty dangerous. For example I think `instance` above may be leaked here, preventing the instance from being dropped? In any case could this be reorganized to:

```rust
let result = {
    // ...
};
if let Err(trap) = result {
    raise_lib_trap(trap);
}
```

with a comment about how the block-structure is required to make sure dtors run?

Comment (alexcrichton) [reply crates/runtime/src/libcalls.rs]: So a possible alternative to this as well might be to do something entirely different wrt traps. For example each of these intrinsic functions could return an i32 which, if zero, would cause a trap. The generated code would then do the test-and-trap itself and have an entry in the trap map for why the trap is happening at that address.

I feel like that may be a bit cleaner than doing the longjmp ourselves, but functionally it's all the same so it doesn't matter too too much. I'm just overly worried about these unwinding functions :(

File: crates/runtime/src/libcalls.rs [hunk 14]
```diff
@@ -137,3 +142,195 @@ pub unsafe extern "C" fn wasmtime_imported_memory32_size(
...
+#[no_mangle]
+pub unsafe extern "C" fn wasmtime_elem_drop(vmctx: *mut VMContext, elem_index: u32) {
+    let elem_index = PassiveElemIndex::from_u32(elem_index);
+    let instance = (&mut *vmctx).instance();
+    instance.elem_drop(elem_index);
+}
+
+/// Implementation of `memory.copy` for locally defined memories.
+#[no_mangle]
+pub unsafe extern "C" fn wasmtime_memory_copy(
```
Comment (alexcrichton) [inline crates/runtime/src/libcalls.rs]: Could this be renamed `wasmtime_defined_memory_copy`?

(or perhaps get merged with `wasmtime_imported_memory_copy`, I'm still not entirely sure why there's two separate functions)

File: crates/runtime/src/table.rs [hunk 15]
```diff
@@ -87,6 +89,56 @@ impl Table {
...
+                },
+                backtrace: Backtrace::new(),
+            });
+        }
+
+        let srcs = src_index..src_index + len;
+        let dsts = dst_index..dst_index + len;
+
+        // Note on the unwraps: the bounds check above means that these will
+        // never panic.
```
Comment (alexcrichton) [inline crates/runtime/src/table.rs]: Mind throwing a TODO here that we should do something "more raw"?

File: tests/wast_testsuites.rs [hunk 16]
```diff
@@ -16,12 +16,29 @@ fn run_wast(wast: &str, strategy: Strategy) -> anyhow::Result<()> {
...
+        || wast.iter().any(|s| s == "elem-ref-null.wast")
+        || wast.iter().any(|s| s == "memory-copy.wast")
+        || wast.iter().any(|s| s == "imported-memory-copy.wast")
+        || wast
+            .iter()
+            .any(|s| s == "table_copy_on_imported_tables.wast");
+
+    // And bulk memory also assumes support for reference types (e.g. multiple
+    // tables).
+    let reftypes = reftypes || bulk_mem;
```
Comment (alexcrichton) [inline tests/wast_testsuites.rs]: These new conditions look a bit odd to me. It looks like `bulk_mem` is on if `reftypes` is enabled, but the `bulk_mem` also then enables `reftypes` later on. I thought the dependency was that `reftypes` implied `bulk_mem`, right? If so that's already done in `Config` I believe, so we just need to know here if bulk memory is enabled.

Oh actually I think to get the right behavior from `Config` you need to do the bulk memory setter first, and then the reference types setter (it's a bit wonky).

Also, given that we have a folder for bulk memory, could all our local `misc_testsuite` tests related to bulk memory be placed in a same-name folder to avoid having to list all tests here?

File: crates/runtime/src/instance.rs (hunk 3 above)
Comment (fitzgen) [reply crates/runtime/src/instance.rs]: I suppose that is technically correct. In `cranelift_wasm` we only ever report the indices for passive elements even though its the same index space. (So consumers get a sparse view of the index space).

Filed this issue in cranelift: https://github.com/bytecodealliance/cranelift/issues/1407

File: crates/runtime/src/instance.rs (hunk 11 above)
Comment (fitzgen) [reply crates/runtime/src/instance.rs]: Good eye, I was focused on chugging away to get the spec tests passing, but this does look like a bug in the spec tests themselves to me as well. I filed https://github.com/WebAssembly/bulk-memory-operations/issues/138 to clarify.

File: crates/runtime/src/libcalls.rs (hunk 13 above)
Comment (fitzgen) [reply crates/runtime/src/libcalls.rs]: Yeah, we would probably want to make `TrapCode` be `repr(C)` and also have a `Ok` variant or something like that for success (or maybe a new `ResultCode` so that `TrapCode` itself doesn't have an `Ok`).

This seems like a good plan, but also something we can do not in this particular PR.

File: crates/runtime/src/libcalls.rs (hunk 14 above)
Comment (fitzgen) [reply crates/runtime/src/libcalls.rs]: mostly because I was cribbing from how we do `memory.size` and `memory.grow`

I'll do the renaming in this PR and file an issue for merging imported/defined memory impls together. https://github.com/bytecodealliance/wasmtime/issues/985

File: tests/wast_testsuites.rs (hunk 16 above)
Comment (fitzgen) [reply tests/wast_testsuites.rs]: > Also, given that we have a folder for bulk memory, could all our local `misc_testsuite` tests related to bulk memory be placed in a same-name folder to avoid having to list all tests here?

FWIW, we don't have a bulk memory directory in our misc test suite but I can make one.

**Comment by fitzgen** [comment]
@alexcrichton I've addressed all your feedback, do you want to take another look at this?

File: crates/runtime/src/instance.rs (hunk 11 above)
Comment (alexcrichton) [reply crates/runtime/src/instance.rs]: Given https://github.com/WebAssembly/bulk-memory-operations/issues/138#issuecomment-591116546 it sounds like this loop doesn't need to change much here?

File: build.rs [hunk 17]
```diff
@@ -136,6 +150,8 @@ fn write_testsuite_tests(
     strategy: &str,
 ) -> anyhow::Result<()> {
     let path = path.as_ref();
+    println!("cargo:rerun-if-changed={}", path.display());
```
[the remark this answers is not in the document]
Comment (fitzgen) [reply build.rs]: I guess we don't need them for the files themselves if we load them at runtime, but its nice to have them on the directory so that if you add a new file, you don't have to touch build.rs

[the remark this answers is not in the document]
Comment (alexcrichton) [reply build.rs]: I personally find mtimes of directories sketchy enough that they're not worth using. It's pretty platform specific what the mtime measures and I think on some platforms it's as bad as printing to rerun on all the contents of the directory. I'd personally prefer to just document the deficiency here and say "touch `build.rs` if you add a new file". We'll always catch it on CI anyway.
